### ISSUE TITLE
COPYs parsed inside a program don't use MultilineScanState

### ISSUE THREAD

**Issue by smedilol** [TypeCobolTeam/TypeCobol#315]
CHILD OF #311 and #235 
With the fix of PR #311, and the the following code:

``` cobol
IDENTIFICATION DIVISION.
PROGRAM-ID. Decimal.
ENVIRONMENT DIVISION.
CONFIGURATION SECTION.
SOURCE-COMPUTER. IBM-370.
special-names. decimal-point is comma.
DATA DIVISION.
working-storage section.
COPY Decimal2.
```

And Decimal2:

``` cobol
 01 RATE PIC 9V99999  VALUE 9,17957. 
```

I get the error:
`[37,41]<27,Error,Syntax>Syntax error : extraneous input '17957' expecting PeriodSeparator (RuleStack=codeElement>dataDescriptionEntry, OffendingSymbol=[37,41:17957]<IntegerLiteral>{17957} in COPY Decimal2 on line 1)`

The problem is that MultilineScanState.DecimalPointIsComma is true when parsing the program but false when parsing the COPY.
The parsing of the COPY doesn't reuse the MultilineScanState of the program.

**Comment by smedilol** [comment]
My quick analysis:
Interface IProcessedTokensDocumentProvider.GetProcessedTokensDocument which seems to be only used to parse copy should evolve to get a MultilineScanState (or an equivalent).

FileCompiler, CompilationDocument and ScannerStep should get this MultilineScanState (or an equivalent). In particular method ScannerStep.ScanDocument.

@laurentprudhon what do you think?

**Comment by laurentprudhon** [comment]
COPY files are often reused many times, in many different programs, and they can be big.
The TypeCobol compiler has been designed to enable real time code analysis in an editor : we can't afford to waste CPU or memory.
The idea was to scan each COPY file only once, cache the result, and then reuse its tokens via iterators in all the programs which include this file.
The problem you uncovered in this issue is that this goal is not possible to achieve without limitations, because scanning is contextual in Cobol. If we include a COPY in two different programs, a scan of the exact same COPY could result in two different token streams.
In practice however, we would have a hard time coming up with a Cobol fragment that would still be valid when interpreted differently in two different scanning contexts.

**Comment by laurentprudhon** [comment]
When we start scanning a file, we initialize the scanner context (MultilineScanState) with the following parameters :
ScanFirstLine(TokensLine tokensLine, bool insideDataDivision, bool decimalPointIsComma, bool withDebuggingMode, Encoding encodingForAlphanumericLiterals, TypeCobolOptions compilerOptions)

The three boolean parameters are initialized to false when we start scanning a Program.

They should be initialized with the values found in MultilineScanState at the point where a COPY import directive is found during the scan of the importing program.

**Comment by laurentprudhon** [comment]
How to enhance the current implementation and fixe this issue :

1. PreprocessorStep, line 233 : pass the ScanState of the current line when calling GetProcessedTokensDocument

ProcessedTokensDocument importedDocumentSource = processedTokensDocumentProvider.GetProcessedTokensDocument(copyDirective.LibraryName, copyDirective.TextName, tokensLineWithCopyDirective.ScanState);

2. CompilationProject, line 181 : pass all the initial scan context properties to the FileCompiler constructor 

FileCompiler fileCompiler = new FileCompiler(libraryName, textName, SourceFileProvider, this, ColumnsLayout, scanState.InsideDataDivision, scanState.DecimalPointIsComma, scanState.WithDebuggingMode, CompilationOptions, null, true);

3. FileCompiler line 133 => CompilationDocument line 387  => ScannerStep line 46 : propagate this initial scan context until this call

Scanner.ScanFirstLine(tokensLine, false, false, false, textSourceInfo.EncodingForAlphanumericLiterals, compilerOptions);

4. Replace the three default "false" values with the initial scan context received from PreprocessorStep if available

**Comment by smedilol** [comment]
Thanks.
I'll also  update the cacheKey in CompilationProject.GetProcessedTokensDocument to store the value of insideDataDivision, decimalPointIsComma and withDebuggingMode.

**Comment by smedilol** [comment]
Fixed by 	f88de78
